x-pack/plugins/observability/server/lib/rules/custom_threshold/lib/create_rate_aggregation.ts

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import moment from 'moment';
+import { calculateRateTimeranges } from '../utils';
+export const createRateAggsBucketScript = (
+  timeframe: { start: number; end: number },
+  id: string
+) => {
+  const { intervalInSeconds } = calculateRateTimeranges({
+    to: timeframe.end,
+    from: timeframe.start,
+  });
+  return {
+    [id]: {
+      bucket_script: {
+        buckets_path: {
+          first: `${id}_first_bucket.maxValue`,
+          second: `${id}_second_bucket.maxValue`,
+        },
+        script: `params.second > 0.0 && params.first > 0.0 && params.second > params.first ? (params.second - params.first) / ${intervalInSeconds}: null`,
+      },
+    },
+  };
+};
+
+export const createRateAggsBuckets = (
+  timeframe: { start: number; end: number },
```

Comment:
@maryam-saeidi can you elaborate more? what do mean by `the timeframe is doubled to cover the first and second buffers correctly`
From the screenshots, I can't see the double difference you mentioned

Reply:
In the first highlighted section, for the currentPeriod, the metric threshold is calculated from min 47 to 49, but in the custom threshold, it is calculated from 48 to 49. As a result, in the custom threshold, each bucket is 30 seconds instead of 1 minute. (both rules have 1 minute in the `for the last` field)

Reply:
[Here](https://github.com/elastic/kibana/blob/main/x-pack/plugins/infra/server/lib/alerting/metric_threshold/lib/create_timerange.ts#L20) is where the min bucket doubled for the rate aggregation in the metric threshold.

Reply:
@maryam-saeidi, good catch!

Reply:
@maryam-saeidi, here is the [fix](https://github.com/elastic/kibana/pull/173728/commits/ad73a88d44503430e9dae9f05c42fcf0a3f50fc1) for the current period section of the query.
Then I noticed that the bucket time range is calculated in a different place, and here is where it [fix](https://github.com/elastic/kibana/pull/173728/commits/d7306218d912e2ef1d67937c5ea5c53fda0dcc15).  
Please give it a try. I tested on my side, and it doubled the interval. Here is the outcome of my test:
```
{
   "index":"high*",
   "allow_no_indices":true,
   "ignore_unavailable":true,
   "body":{
      "track_total_hits":true,
      "query":{
         "bool":{
            "filter":[
               {
                  "range":{
                     "@timestamp":{
                        "gte":"2024-01-04T15:57:14.392Z",
                        "lte":"2024-01-04T16:00:17.506Z"
                     }
                  }
               }
            ]
         }
      },
      "size":0,
      "aggs":{
         "all":{
            "filters":{
               "filters":{
                  "all":{
                     "match_all":{
                        
                     }
                  }
               }
            },
            "aggs":{
               "currentPeriod":{
                  "filters":{
                     "filters":{
                        "all":{
                           "range":{
                              "@timestamp":{
                                 "gte":"2024-01-04T15:58:17.506Z",
                                 "lte":"2024-01-04T16:00:17.506Z"
                              }
                           }
                        }
                     }
                  },
                  "aggs":{
                     "aggregatedValue_A_first_bucket":{
                        "filter":{
                           "range":{
                              "@timestamp":{
                                 "gte":"2024-01-04T15:58:17.506Z",
                                 "lt":"2024-01-04T15:59:17.506Z"
                              }
                           }
                        },
                        "aggs":{
                           "maxValue":{
                              "max":{
                                 "field":"system.network.in.bytes"
                              }
                           }
                        }
                     },
                     "aggregatedValue_A_second_bucket":{
                        "filter":{
                           "range":{
                              "@timestamp":{
                                 "gte":"2024-01-04T15:59:17.506Z",
                                 "lt":"2024-01-04T16:00:17.506Z"
                              }
                           }
                        },
                        "aggs":{
                           "maxValue":{
                              "max":{
                                 "field":"system.network.in.bytes"
                              }
                           }
                        }
                     },
                     "aggregatedValue_A":{
                        "bucket_script":{
                           "buckets_path":{
                              "first":"aggregatedValue_A_first_bucket.maxValue",
                              "second":"aggregatedValue_A_second_bucket.maxValue"
                           },
                           "script":"params.second > 0.0 && params.first > 0.0 && params.second > params.first ? (params.second - params.first) / 60: null"
                        }
                     },
                     "aggregatedValue":{
                        "bucket_script":{
                           "buckets_path":{
                              "A":"aggregatedValue_A"
                           },
                           "script":{
                              "source":"params.A",
                              "lang":"painless"
                           }
                        }
                     }
                  }
               },
               "shouldTrigger":{
                  "bucket_script":{
                     "buckets_path":{
                        "value":"currentPeriod['all']>aggregatedValue"
                     },
                     "script":{
                        "source":"params.value > params.threshold ? 1 : 0",
                        "params":{
                           "threshold":100
                        }
                     }
                  }
               }
            }
         }
      }
   }
}
```